### PR TITLE
Improving sidebar item behavior

### PR DESCRIPTION
**Pull request by Augmentador**

### CODE REVIEW

File: lib/src/layout/sidebar/sidebar_item.dart [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import 'package:flutter/foundation.dart';
-import 'package:macos_ui/macos_ui.dart';
-import 'package:macos_ui/src/library.dart';
+
+import '../../../macos_ui.dart';
+import '../../library.dart';
```
Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_item.dart]: This is just noise, presumably frome something in your IDE setup automatically converting package to relative imports. Please revert this. In VSCode it should be possible to set per-project settings to disable whatever is making this change just for this repo.

File: lib/src/layout/sidebar/sidebar_items.dart [hunk 2]
```diff
@@ -1,5 +1,5 @@
-import 'package:macos_ui/macos_ui.dart';
-import 'package:macos_ui/src/library.dart';
+import '../../../macos_ui.dart';
+import '../../library.dart';
```
Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart]: Same, see previous comment.

File: lib/src/layout/sidebar/sidebar_items.dart [hunk 3]
```diff
@@ -47,26 +47,24 @@ class SidebarItems extends StatelessWidget {
...
+  });
 
   /// The [SidebarItem]s used by the sidebar. If no items are provided,
   /// the sidebar is not rendered.
   final List<SidebarItem> items;
 
-  /// The current selected index. It must be in the range of 0 to
-  /// [items.length]
-  final int currentIndex;
+  final String? currenIdentifier;
```
Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart]: Why the change from indexes to unique string ids? I don't necessarily have a problem with it, but if we want to get these changes upstreamed there needs to be a reason for changing the API.

File: example/lib/main.dart [hunk 4]
```diff
@@ -173,82 +174,57 @@ class _WidgetGalleryState extends State<WidgetGallery> {
...
           builder: (context, _) {
             return SidebarItems(
-              currentIndex: pageIndex,
-              onChanged: (i) => setState(() => pageIndex = i),
-              scrollController: ScrollController(),
+              currenIdentifier: currentIdentifier,
+              onChanged: (id) => setState(() {
+                currentIdentifier = id;
+              }),
+              scrollController: controller,
```
Comment (jmatth) [inline example/lib/main.dart]: The example fails to compile because this constructor argument does not exist. Additionally, there's something weird going on in your git history: this branch is forked off of `cc-custom`, but then merges that branch back into itself after one commit. I'm guessing the two are related.

**Review by jmatth — CHANGES_REQUESTED** [review]
This may just be an unintentional side effect of changing the sidebar for testing, but it is no longer possible to change pages in the example app. Every sidebar item displays the same content in the main area. For testing purposes, let's leave the current sidebar items in the example app alone and just add a nested hierarchy to the bottom of the list for testing / demonstration purposes.

Something that wasn't in the original issue but is apparent from testing: expanding/collapsing a parent should only happen when the user clicks the chevron. This matches native behavior and avoids the annoying issue of a section closing when you try to switch to it.

File: lib/src/layout/sidebar/sidebar_item.dart (hunk 1 above)
Comment (Augmentador) [reply lib/src/layout/sidebar/sidebar_item.dart]: Apologies. I have to turn this stuff off 🤦

File: lib/src/layout/sidebar/sidebar_items.dart (hunk 3 above)
Comment (Augmentador) [reply lib/src/layout/sidebar/sidebar_items.dart]: Previously the use of `int` was due to the fact that the package only kept track of the index of the item, and given that it could only ever be one level deep the structure worked as follow
- Regular Item (1)
- Folder
   - Regular Item (2)
   - Regular Item (3)
- Regular Item (4)

This breaks down with multiple nested objects so a unique identifier for each is necessary. In this instance i made the use of string as they seemed for more fitting. Now we can allow for (n) nested items as follow:

- Regular Item (id)
- Folder
   - Regular Item (id)
   - Regular Item (id)
   - Folder
        - Regular Item (id)
        - Regular Item (id)
- Regular Item (id)

Comment (jmatth) [reply lib/src/layout/sidebar/sidebar_items.dart]: It should be possible to produce the same result with a preorder traversal of the items:

- Regular Item (1)
- Folder (2)
  - Regular Item (3)
  - Regular Item (4)
  - Folder (5)
    - Regular Item (6)
    - Regular Item (7)
- Regular Item (8)

Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart]: Typo, you're missing a `t`. Should be `currentIdentifier`.

Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart]: Please add a docstring similar to the one on the old `currentIndex` variable. Something like:

```
The id of the currently selected item. There must be a [SidebarItem] with a matching id in [items].
```

Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart]: Should this be nullable? The current API requires a sidebar item to be selected at all times. I'm inclined to stick with that for now.

File: lib/src/layout/sidebar/sidebar_items.dart [hunk 5]
```diff
@@ -47,26 +47,24 @@ class SidebarItems extends StatelessWidget {
...
 
   /// Called when the current selected index should be changed.
```
Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart]: Change `index` to `identifier` in the docstring here. Probably worth doing a quick find/replace through the file to catch any other references to index as well.

File: lib/src/layout/sidebar/sidebar_items.dart [hunk 6]
```diff
@@ -190,16 +187,15 @@ class _SidebarItem extends StatelessWidget {
...
   }) : super(key: key);
 
   /// The widget to lay out first.
   ///
   /// Typically an [Icon]
   final SidebarItem item;
 
-  /// Whether the item is selected or not
-  final bool selected;
+  final String? currenIdentifier;
```
Comment (jmatth) [inline lib/src/layout/sidebar/sidebar_items.dart]: Why the change from a boolean to the current identifier here? Seems like the selected state could still be computed where this class is instantiated.

**Review by jmatth — CHANGES_REQUESTED** [review]

File: lib/src/layout/sidebar/sidebar_items.dart (hunk 6 above)
Comment (Augmentador) [reply lib/src/layout/sidebar/sidebar_items.dart]: Setting this to an identifier allows us to infer the selection state at any level of the sidebar items tree.

**Review by jmatth — APPROVED** [review]
